File: obs_mcp/scene_items.py
#!/usr/bin/env python3

from typing import Any, Dict, List, Optional, Union

from .client import obs_client
from .server import mcp

@mcp.tool()
async def get_scene_item_list(scene_name: str) -> List[Dict[str, Any]]:
    """
    Gets a list of all scene items in a scene.
    
    Args:
        scene_name: Name of the scene to get the items of
    
    Returns:
        List of scene items (each with sceneItemId, sourceName, sourceKind, sceneItemIndex)
    """
    response = await obs_client.send_request("GetSceneItemList", {"sceneName": scene_name})
    return response.get("sceneItems", [])

@mcp.tool()
async def get_group_item_list(scene_name: str, group_name: str) -> List[Dict[str, Any]]:
    """
    Gets a list of all scene items in a group.
    
    Args:
        scene_name: Name of the scene the group is in
        group_name: Name of the group to get the items of
    
    Returns:
        List of scene items (each with sceneItemId, sourceName, sourceKind, sceneItemIndex)
    """
    response = await obs_client.send_request("GetGroupSceneItemList", {
        "sceneName": scene_name,
        "groupName": group_name
    })
    return response.get("sceneItems", [])

@mcp.tool()
async def create_scene_item(scene_name: str, source_name: str, enabled: bool = True) -> int:
    """
    Creates a new scene item in a scene.
    
    Args:
        scene_name: Name of the scene to create the item in
        source_name: Name of the source to add to the scene
        enabled: Whether to set the scene item to enabled or disabled
    
    Returns:
        ID of the created scene item
    """
    response = await obs_client.send_request("CreateSceneItem", {
        "sceneName": scene_name,
        "sourceName": source_name,
        "sceneItemEnabled": enabled
    })
    return response.get("sceneItemId", 0)

@mcp.tool()
async def remove_scene_item(scene_name: str, scene_item_id: int) -> None:
    """
    Removes a scene item from a scene.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item to remove
    """
    await obs_client.send_request("RemoveSceneItem", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id
    })

@mcp.tool()
async def duplicate_scene_item(scene_name: str, scene_item_id: int, 
                              destination_scene_name: Optional[str] = None) -> int:
    """
    Duplicates a scene item in a scene.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item to duplicate
        destination_scene_name: Name of the scene to create the duplicated item in (defaults to original scene)
    
    Returns:
        ID of the duplicated scene item
    """
    payload = {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id
    }
    
    if destination_scene_name:
        payload["destinationSceneName"] = destination_scene_name
    
    response = await obs_client.send_request("DuplicateSceneItem", payload)
    return response.get("sceneItemId", 0)

@mcp.tool()
async def get_scene_item_id(scene_name: str, source_name: str, search_offset: int = 0) -> int:
    """
    Gets the ID of a scene item in a scene.
    
    Args:
        scene_name: Name of the scene the item is in
        source_name: Name of the source to find the ID of
        search_offset: Number of matches to skip
    
    Returns:
        ID of the scene item
    """
    response = await obs_client.send_request("GetSceneItemId", {
        "sceneName": scene_name,
        "sourceName": source_name,
        "searchOffset": search_offset
    })
    return response.get("sceneItemId", 0)

@mcp.tool()
async def get_scene_item_enabled(scene_name: str, scene_item_id: int) -> bool:
    """
    Gets the enabled state of a scene item.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
    
    Returns:
        Whether the scene item is enabled
    """
    response = await obs_client.send_request("GetSceneItemEnabled", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id
    })
    return response.get("sceneItemEnabled", False)

@mcp.tool()
async def set_scene_item_enabled(scene_name: str, scene_item_id: int, enabled: bool) -> None:
    """
    Sets the enabled state of a scene item.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
        enabled: New enabled state of the scene item
    """
    await obs_client.send_request("SetSceneItemEnabled", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id,
        "sceneItemEnabled": enabled
    })

@mcp.tool()
async def get_scene_item_locked(scene_name: str, scene_item_id: int) -> bool:
    """
    Gets the locked state of a scene item.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
    
    Returns:
        Whether the scene item is locked
    """
    response = await obs_client.send_request("GetSceneItemLocked", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id
    })
    return response.get("sceneItemLocked", False)

@mcp.tool()
async def set_scene_item_locked(scene_name: str, scene_item_id: int, locked: bool) -> None:
    """
    Sets the locked state of a scene item.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
        locked: New locked state of the scene item
    """
    await obs_client.send_request("SetSceneItemLocked", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id,
        "sceneItemLocked": locked
    })

@mcp.tool()
async def get_scene_item_index(scene_name: str, scene_item_id: int) -> int:
    """
    Gets the index position of a scene item in a scene.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
    
    Returns:
        Index position of the scene item
    """
    response = await obs_client.send_request("GetSceneItemIndex", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id
    })
    return response.get("sceneItemIndex", 0)

@mcp.tool()
async def set_scene_item_index(scene_name: str, scene_item_id: int, index: int) -> None:
    """
    Sets the index position of a scene item in a scene.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
        index: New index position for the scene item
    """
    await obs_client.send_request("SetSceneItemIndex", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id,
        "sceneItemIndex": index
    })

@mcp.tool()
async def get_scene_item_transform(scene_name: str, scene_item_id: int) -> Dict[str, Any]:
    """
    Gets the transform/crop info of a scene item.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
    
    Returns:
        Dict with transform information (position, rotation, scale, crop, bounds)
    """
    return await obs_client.send_request("GetSceneItemTransform", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id
    })

@mcp.tool()
async def set_scene_item_transform(scene_name: str, scene_item_id: int, transform: Dict[str, Any]) -> None:
    """
    Sets the transform/crop info of a scene item.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
        transform: Dict with transform properties to set
    """
    await obs_client.send_request("SetSceneItemTransform", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id,
        "sceneItemTransform": transform
    })

@mcp.tool()
async def get_scene_item_blend_mode(scene_name: str, scene_item_id: int) -> str:
    """
    Gets the blend mode of a scene item.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
    
    Returns:
        Current blend mode of the scene item
    """
    response = await obs_client.send_request("GetSceneItemBlendMode", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id
    })
    return response.get("sceneItemBlendMode", "")

@mcp.tool()
async def set_scene_item_blend_mode(scene_name: str, scene_item_id: int, blend_mode: str) -> None:
    """
    Sets the blend mode of a scene item.
    
    Args:
        scene_name: Name of the scene the item is in
        scene_item_id: ID of the scene item
        blend_mode: New blend mode (OBS_BLEND_NORMAL, etc.)
    """
    await obs_client.send_request("SetSceneItemBlendMode", {
        "sceneName": scene_name,
        "sceneItemId": scene_item_id,
        "sceneItemBlendMode": blend_mode
    })

File: src/tools/scene-items.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { OBSWebSocketClient } from "../client.js";
import { z } from "zod";

export async function initialize(server: McpServer, client: OBSWebSocketClient): Promise<void> {
  // GetSceneItemList tool
  server.tool(
    "obs-get-scene-items",
    "Get a list of all scene items in a scene",
    {
      sceneName: z.string().describe("The name of the scene to get items from")
    },
    async ({ sceneName }) => {
      try {
        const sceneItems = await client.sendRequest("GetSceneItemList", { sceneName });
        return {
          content: [
            {
              type: "text",
              text: JSON.stringify(sceneItems, null, 2)
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting scene items: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // CreateSceneItem tool
  server.tool(
    "obs-create-scene-item",
    "Create a scene item for a source in a scene",
    {
      sceneName: z.string().describe("The scene to add the source to"),
      sourceName: z.string().describe("The name of the source to add"),
      enabled: z.boolean().optional().describe("Whether the scene item is enabled/visible (default: true)")
    },
    async ({ sceneName, sourceName, enabled = true }) => {
      try {
        const response = await client.sendRequest("CreateSceneItem", {
          sceneName,
          sourceName,
          sceneItemEnabled: enabled
        });

        return {
          content: [
            {
              type: "text",
              text: `Successfully added ${sourceName} to ${sceneName} with ID: ${response.sceneItemId}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error creating scene item: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // RemoveSceneItem tool
  server.tool(
    "obs-remove-scene-item",
    "Remove a scene item from a scene",
    {
      sceneName: z.string().describe("The scene to remove the item from"),
      sceneItemId: z.number().describe("The ID of the scene item to remove")
    },
    async ({ sceneName, sceneItemId }) => {
      try {
        await client.sendRequest("RemoveSceneItem", { sceneName, sceneItemId });

        return {
          content: [
            {
              type: "text",
              text: `Successfully removed item with ID ${sceneItemId} from ${sceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error removing scene item: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // SetSceneItemEnabled tool
  server.tool(
    "obs-set-scene-item-enabled",
    "Show or hide a scene item",
    {
      sceneName: z.string().describe("The scene that the source belongs to"),
      sceneItemId: z.number().describe("The ID of the scene item"),
      enabled: z.boolean().describe("Whether to show (true) or hide (false) the item")
    },
    async ({ sceneName, sceneItemId, enabled }) => {
      try {
        await client.sendRequest("SetSceneItemEnabled", {
          sceneName,
          sceneItemId,
          sceneItemEnabled: enabled
        });

        return {
          content: [
            {
              type: "text",
              text: `Successfully ${enabled ? "showed" : "hid"} item with ID ${sceneItemId} in ${sceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error setting scene item visibility: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // GetSceneItemTransform tool
  server.tool(
    "obs-get-scene-item-transform",
    "Get the position, rotation, scale, or crop of a scene item",
    {
      sceneName: z.string().describe("The scene the item is in"),
      sceneItemId: z.number().describe("The ID of the scene item")
    },
    async ({ sceneName, sceneItemId }) => {
      try {
        const response = await client.sendRequest("GetSceneItemTransform", { sceneName, sceneItemId });
        return {
          content: [
            {
              type: "text",
              text: JSON.stringify(response, null, 2)
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting scene item transform: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // SetSceneItemTransform tool
  server.tool(
    "obs-set-scene-item-transform",
    "Set the position, rotation, scale, or crop of a scene item",
    {
      sceneName: z.string().describe("The scene the item is in"),
      sceneItemId: z.number().describe("The ID of the scene item"),
      positionX: z.number().optional().describe("The x position"),
      positionY: z.number().optional().describe("The y position"),
      rotation: z.number().optional().describe("The rotation in degrees"),
      scaleX: z.number().optional().describe("The x scale factor"),
      scaleY: z.number().optional().describe("The y scale factor"),
      cropTop: z.number().optional().describe("The number of pixels cropped off the top"),
      cropBottom: z.number().optional().describe("The number of pixels cropped off the bottom"),
      cropLeft: z.number().optional().describe("The number of pixels cropped off the left"),
      cropRight: z.number().optional().describe("The number of pixels cropped off the right")
    },
    async (params) => {
      try {
        const { sceneName, sceneItemId, ...transformParams } = params;

        // Build the transform object
        const sceneItemTransform: Record<string, any> = {};

        if (transformParams.positionX !== undefined || transformParams.positionY !== undefined) {
          sceneItemTransform.positionX = transformParams.positionX;
          sceneItemTransform.positionY = transformParams.positionY;
        }

        if (transformParams.rotation !== undefined) {
          sceneItemTransform.rotation = transformParams.rotation;
        }

        if (transformParams.scaleX !== undefined || transformParams.scaleY !== undefined) {
          sceneItemTransform.scaleX = transformParams.scaleX;
          sceneItemTransform.scaleY = transformParams.scaleY;
        }

        if (transformParams.cropTop !== undefined || transformParams.cropBottom !== undefined ||
          transformParams.cropLeft !== undefined || transformParams.cropRight !== undefined) {
          if (transformParams.cropTop !== undefined) sceneItemTransform.cropTop = transformParams.cropTop;
          if (transformParams.cropBottom !== undefined) sceneItemTransform.cropBottom = transformParams.cropBottom;
          if (transformParams.cropLeft !== undefined) sceneItemTransform.cropLeft = transformParams.cropLeft;
          if (transformParams.cropRight !== undefined) sceneItemTransform.cropRight = transformParams.cropRight;
        }

        await client.sendRequest("SetSceneItemTransform", {
          sceneName,
          sceneItemId,
          sceneItemTransform
        });

        return {
          content: [
            {
              type: "text",
              text: `Successfully updated transform for item with ID ${sceneItemId} in ${sceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error setting scene item transform: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // GetSceneItemIdByName tool
  server.tool(
    "obs-get-scene-item-id",
    "Get the ID of a scene item by its source name",
    {
      sceneName: z.string().describe("The scene name to search in"),
      sourceName: z.string().describe("The source name to find")
    },
    async ({ sceneName, sourceName }) => {
      try {
        const response = await client.sendRequest("GetSceneItemId", {
          sceneName,
          sourceName
        });

        return {
          content: [
            {
              type: "text",
              text: `Scene item ID for ${sourceName} in ${sceneName}: ${response.sceneItemId}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting scene item ID: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );
}